contrib/trace_propagators/src/main/java/io/opentelemetry/contrib/trace/propagation/AwsXRayPropagator.java

```diff
@@ -0,0 +1,264 @@
+    return FIELDS;
+  }
+
+  @Override
+  public <C> void inject(Context context, @Nullable C carrier, Setter<C> setter) {
+    Objects.requireNonNull(context, "context");
+    Objects.requireNonNull(setter, "setter");
+
+    Span span = TracingContextUtils.getSpan(context);
+    if (!span.getContext().isValid()) {
+      return;
+    }
+
+    SpanContext spanContext = span.getContext();
+
+    String otTraceId = spanContext.getTraceId().toLowerBase16();
+    String xrayTraceId =
+        TRACE_ID_VERSION
+            + TRACE_ID_DELIMITER
+            + otTraceId.substring(0, TRACE_ID_FIRST_PART_LENGTH)
+            + TRACE_ID_DELIMITER
+            + otTraceId.substring(TRACE_ID_FIRST_PART_LENGTH);
+    String parentId = spanContext.getSpanId().toLowerBase16();
+    char samplingFlag = spanContext.getTraceFlags().isSampled() ? IS_SAMPLED : NOT_SAMPLED;
+    // TODO: Add OT trace state to the X-Ray trace header
+
+    String traceHeader =
+        TRACE_ID_KEY
+            + KV_DELIMITER
+            + xrayTraceId
+            + TRACE_HEADER_DELIMITER
+            + PARENT_ID_KEY
+            + KV_DELIMITER
+            + parentId
+            + TRACE_HEADER_DELIMITER
+            + SAMPLED_FLAG_KEY
+            + KV_DELIMITER
+            + samplingFlag;
+    setter.set(carrier, TRACE_HEADER_KEY, traceHeader);
+  }
+
+  @Override
+  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
+    Objects.requireNonNull(carrier, "carrier");
+    Objects.requireNonNull(getter, "getter");
+
+    SpanContext spanContext = getSpanContextFromHeader(carrier, getter);
+    return TracingContextUtils.withSpan(DefaultSpan.create(spanContext), context);
+  }
+
+  private static <C> SpanContext getSpanContextFromHeader(C carrier, Getter<C> getter) {
+    String traceHeader = getter.get(carrier, TRACE_HEADER_KEY);
+    if (traceHeader == null || traceHeader.isEmpty()) {
+      return SpanContext.getInvalid();
+    }
+
+    TraceId traceId = TraceId.getInvalid();
+    SpanId spanId = SpanId.getInvalid();
+    TraceFlags traceFlags = TraceFlags.getDefault();
+
+    int pos = 0;
+    while (pos < traceHeader.length()) {
+      int delimiterIndex = traceHeader.indexOf(TRACE_HEADER_DELIMITER, pos);
+      final String part;
+      if (delimiterIndex >= 0) {
+        part = traceHeader.substring(pos, delimiterIndex);
+        pos = delimiterIndex + 1;
+      } else {
+        // Last part.
+        part = traceHeader.substring(pos);
+        pos = traceHeader.length();
+      }
+      String trimmedPart = part.trim();
+      int equalsIndex = trimmedPart.indexOf(KV_DELIMITER);
+      if (equalsIndex < 0) {
+        logger.info(
+            "Error parsing X-Ray trace header. Invalid key value pair: "
+                + part
+                + " Returning INVALID span context.");
+        return SpanContext.getInvalid();
+      }
+
+      String value = trimmedPart.substring(equalsIndex + 1);
+
+      if (trimmedPart.startsWith(TRACE_ID_KEY)) {
```

Comment:
Once the parse for each section is done, we can try to check the value and stop immediately if the value is invalid, instead of parsing all sections and then checking them one by one later. Thoughts?

Reply:
@anuraaga